File: src/entity/bullet/Bullet.java
package entity.bullet;

import static gui.Game.TILE_SIZE;

import entity.Entity;
import entity.enemy.Enemy;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import javafx.scene.transform.Rotate;

public abstract class Bullet extends Entity {
	
	int damage;
	Enemy enemy;
	
	double speed, rotationAngle;
	
	Image img;
	Image flames;
	
	public Bullet(double x, double y, Enemy enemy, double speed, int damage) {
		super(x, y);
		this.enemy = enemy;
		this.speed = speed;
		this.damage = damage;
		
		double dx = enemy.getX() - getX();
		double dy = enemy.getY() - getY();
		
		double angle = Math.atan2(dy, dx);
		
		rotationAngle = Math.toDegrees(angle) + 90;
	}

	@Override
	public void draw(GraphicsContext gc) {
		
		double xi = getX() - TILE_SIZE / 2;
		double yi = getY() - TILE_SIZE / 2;
		
		double dy = (double) TILE_SIZE * .40 + Math.sin(Math.toRadians(rotationAngle + 180));
		
		Rotate r;
		
		if (flames != null) {
			r = new Rotate(rotationAngle + 180, getX(), getY());
			gc.save();
			gc.setTransform(r.getMxx(), r.getMyx(), r.getMxy(), r.getMyy(), r.getTx(), r.getTy());
			gc.drawImage(flames, xi, yi - dy, TILE_SIZE, TILE_SIZE);
			gc.restore();
		}
			
		r = new Rotate(rotationAngle, getX(), getY());
		gc.save();
		gc.setTransform(r.getMxx(), r.getMyx(), r.getMxy(), r.getMyy(), r.getTx(), r.getTy());
		gc.drawImage(img, xi, yi, TILE_SIZE, TILE_SIZE);
		gc.restore();
	}
	
	public void move() {
		
		if (!enemy.isDead()) {
			double dx = enemy.getX() - getX();
			double dy = enemy.getY() - getY();
			
			double angle = Math.atan2(dy, dx);
			
			rotationAngle = Math.toDegrees(angle) + 90;
			
			setX(getX() + speed * Math.cos(angle));
			setY(getY() + speed * Math.sin(angle));
		} else {
			double angle = Math.toRadians(rotationAngle - 90);
			
			setX(getX() + speed * Math.cos(angle));
			setY(getY() + speed * Math.sin(angle));
		}
		
	}
	
	public void doDamage() {
		enemy.takeDamage(damage);
	}
	
	public boolean collided() {
		
		if (enemy.isDead()) {
			return false;
		}
		
		return distance(enemy) <= 5;
	}

	public boolean offScreen() {
		return (getX() < -TILE_SIZE || getX() > TILE_SIZE * 17 || getY() < -TILE_SIZE || getY() > 17 * TILE_SIZE);
	}
	
}

File: src/entity/bullet/Bullet.kt
package entity.bullet

import entity.Entity
import entity.enemy.Enemy
import gui.Game.Companion.TILE_SIZE
import javafx.scene.canvas.GraphicsContext
import javafx.scene.image.Image
import javafx.scene.transform.Rotate

abstract class Bullet(x: Double, y: Double, internal var enemy: Enemy, internal var speed: Double, internal var damage: Int) : Entity(x, y) {
    internal var rotationAngle: Double = 0.0

    internal abstract var img: Image
    internal var flames: Image? = null

    init {

        val dx = enemy.x - x
        val dy = enemy.y - y

        val angle = Math.atan2(dy, dx)

        rotationAngle = Math.toDegrees(angle) + 90
    }

    override fun draw(gc: GraphicsContext) {

        val xi = x - TILE_SIZE / 2
        val yi = y - TILE_SIZE / 2

        val dy = TILE_SIZE.toDouble() * .40 + Math.sin(Math.toRadians(rotationAngle + 180))

        var r: Rotate

        if (flames != null) {
            r = Rotate(rotationAngle + 180, x, y)
            gc.save()
            gc.setTransform(r.mxx, r.myx, r.mxy, r.myy, r.tx, r.ty)
            gc.drawImage(flames, xi, yi - dy, TILE_SIZE.toDouble(), TILE_SIZE.toDouble())
            gc.restore()
        }

        r = Rotate(rotationAngle, x, y)
        gc.save()
        gc.setTransform(r.mxx, r.myx, r.mxy, r.myy, r.tx, r.ty)
        gc.drawImage(img, xi, yi, TILE_SIZE.toDouble(), TILE_SIZE.toDouble())
        gc.restore()
    }

    open fun move() {

        if (!enemy.isDead) {
            val dx = enemy.x - x
            val dy = enemy.y - y

            val angle = Math.atan2(dy, dx)

            rotationAngle = Math.toDegrees(angle) + 90

            x += speed * Math.cos(angle)
            y += speed * Math.sin(angle)
        } else {
            val angle = Math.toRadians(rotationAngle - 90)

            x += speed * Math.cos(angle)
            y += speed * Math.sin(angle)
        }

    }

    fun doDamage() {
        enemy.takeDamage(damage)
    }

    fun collided(): Boolean {

        if (enemy.isDead) {
            return false
        }

        return distance(enemy) <= 5
    }

    fun offScreen(): Boolean {
        return x < -TILE_SIZE || x > TILE_SIZE * 17 || y < -TILE_SIZE || y > 17 * TILE_SIZE
    }

}
